kernel: smp: Optimize delivery of IPIs

peter-mitsis opened the pull request:
This set of commits is a step towards optimising the delivery of inter-processor interrupts (IPIs). It does this by attempting to deliver IPIs only to relevant CPUs.

The bulk of these changes revolve around three ideas.

1. The kernel's _pending_ipi_ field has been tweaked from being a single flag that indicates IPIs are needed to a bitmap whose  bits indicate the CPUs that need to receive an IPI.
2. When a thread is made ready, only CPUs that are executing a lower priority thread get flagged as needing an IPI as they are the only ones that have a chance to schedule the newly readied thread.
3. The _arch_sched_ipi()_ API now accepts a parameter identifying which CPUs should receive an IPI. If the platform allows it, a targeted IPI is sent. If not, then the IPIs are still broadcast to all CPUs.

peter-mitsis committed:
kernel: Add CONFIG_IPI_OPTIMIZE

The CONFIG_IPI_OPTIMIZE configuration option allows for the flagging
and subsequent signaling of IPIs to be optimized.

It does this by making each bit in the kernel's pending_ipi field
a flag that indicates whether the corresponding CPU might need an IPI
to trigger the scheduling of a new thread on that CPU.

When a new thread is made ready, we compare that thread against each
of the threads currently executing on the other CPUs. If there is a
chance that that thread should preempt the thread on the other CPU
then we flag that an IPI is needed for that CPU. That is, a clear bit
indicates that the CPU absolutely will not need to reschedule, while a
set bit indicates that the target CPU must make that determination for
itself.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
kernel: Update IPI usage in k_thread_priority_set()

1. The flagging of IPIs is moved out of k_thread_priority_set() into
z_thread_prio_set(). This allows for an IPI to be done for a thread
that had its priority bumped due to the handling of priority
inheritance from a mutex.

2. k_thread_priority_set()'s check for sched_locked only applies to
non-SMP builds that are using the old arch_swap() framework to switch
between threads.

Incidentally, nearly all calls to flag_ipi() are now performed with
sched_spinlock being locked. The only exception is in slice_timeout().

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
kernel: Add CONFIG_ARCH_HAS_DIRECTED_IPIS

Platforms that support IPIs allow them to be broadcast via the
new arch_sched_broadcast_ipi() routine (replacing arch_sched_ipi()).
Those that also allow IPIs to be directed to specific CPUs may
use arch_sched_directed_ipi() to do so.

As the kernel has the capability to track which CPUs may need an IPI
(see CONFIG_IPI_OPTIMIZE), this commit updates the signalling of
tracked IPIs to use the directed version if supported; otherwise
they continue to use the broadcast version.

Platforms that allow directed IPIs may see a significant reduction
in the number of IPI related ISRs when CONFIG_IPI_OPTIMIZE is
enabled and the number of CPUs increases.  These platforms can be
identified by the Kconfig option CONFIG_ARCH_HAS_DIRECTED_IPIS.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
manifest: optional: Update revision for arch_sched_ipi()

Updates the sof revision. This is needed for it to pick up a
change to the arch_sched_ipi() which has been renamed to
arch_sched_broadcast_ipi().

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

cfriedt commented:
I wonder, if it might be necessary to look at the concept of irq affinity (not just for IPI) in general.

Is that something we support already? Is it something that could be easily added?

Likely not in this PR, if so, just kind of tossing the idea around.

It might be a slightly more general solution (assuming IPI are unique) but could also be useful as a means of dedicating one set of CPUs to I/O and another set for e.g. compute. Maybe we already have this in some way - per-core interrupt masks maybe?

andyross reviewed (commented):
So, the "parallel scheduler implementation" logic in ipi_mask_create() scares me.  That kind of "here's what the other code will run when it does" trick is fragile.  The idea behind broadcast IPIs is that they're simple to understand and prove (and even then there were some historically scary bugs).  Not opposed to optimizing them in principle but IMHO we should work at a single scheduler decision API that provides the logic for both the produce and consume sides of an IPI.

Also: for sure this needs a test, and probably an arch flag like CONFIG_ARCH_HAS_DIRECTED_IPI or something to gate the implementation (right now we pay the cost even if the hardware/arch can't do it).

Finally it would be a shame to add this without getting the arch folks on board for everyone that has support.  x86 for sure wants an implementation (both because it's an Intel submission and because I personally know it can do it), probably ARM64 too.  Looks like riscv is good to go, and ARC just has a missing test to add?

TaiJuWu commented:
I only have a question,
How do you test your PR work well?

When I develop a new feature, I always afraid I miss something or there is logical error.

TaiJuWu commented on arch/riscv/core/smp.c:
```diff
@@ -86,14 +86,16 @@ static atomic_val_t cpu_pending_ipi[CONFIG_MP_MAX_NUM_CPUS];
 
-	for (unsigned int i = 0; i < num_cpus; i++) {
-		if (i != id && _kernel.cpus[i].arch.online) {
+	for (unsigned int i = 0; i < num_cpus; i++, bit <<= 1) {
```
Is uint32_t better?

peter-mitsis replied:
I just left the type unchanged figuring that even if it should be, it was probably outside the scope of what I wanted to change. Aside: I may be wrong but I think that all of our currently supported architectures have an equivalency between 'unsigned int' and 'uint32_t'.

marc-hb replied:
> Is uint32_t better?

If it were, then it would be what `arch_num_cpus()` returns?

Fixed-width types are great for protocols and interacting with hardware but why should they be used in small `for` loops? This could confuse the reader into thinking the width matters here.

I didn't find any https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html about this but maybe I missed one.

andyross replied:
FWIW for a while Zephyr was trying to get serious about MISRA compliance, which disallows variable size types like int and long and requires the stdint explicit ones.  Not sure what the policy is these days.  The existing code remains sort of a mix depending on the specific archaeology of the subsystem.

marc-hb replied:
> >  I didn't find any https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html about this but maybe I missed one.

> FWIW for a while Zephyr was trying to get serious about MISRA compliance,  ...

For anyone who didn't open it, that page is mostly MISRA policies.

peter-mitsis commented on arch/arc/core/smp.c:
```diff
@@ -131,16 +131,17 @@ static void sched_ipi_handler(const void *unused)
 
-	for (i = 0U; i < num_cpus; i++) {
+	for (i = 0U; i < num_cpus; i++, bit <<= 1) {
 		z_arc_connect_ici_generate(i);
```
Will fix.

peter-mitsis commented on arch/x86/core/intel64/smp.c:
```diff
@@ -34,8 +34,10 @@ int arch_smp_init(void)
 {
+	ARG_UNUSED(cpu_bitmap);
+
 	z_loapic_ipi(0, LOAPIC_ICR_IPI_OTHERS, CONFIG_SCHED_IPI_VECTOR);
```
I was originally thinking that that would be outside the scope of this set of changes. I am glad to learn that it is also possible.

peter-mitsis commented on kernel/sched.c:
```diff
@@ -397,15 +399,43 @@ static void move_thread_to_end_of_prio_q(struct k_thread *thread)
+
+		cpu_thread = _kernel.cpus[i].current;
+		if ((cpu_thread != NULL) &&
+		    (cpu_thread->base.prio > thread->base.prio)) {
```
Nice catch! I had not considered deadlines. Using z_sched_prio_cmp() may be a better choice over a straight priority comparison, but I will have to mull this over some more with an eye for races with deadlines (such musings at this time are night are prone to error).

Ignoring the inclusion of deadlines for the moment (only because I haven't given that enough thought yet and it is about 11 PM where I am), I am quite sure that this does not introduce a new scenario where something could happen in the unlocked interim that could confuse things and require an IPI to be delivered. Aside: this does not fix the issue described in #69005, but I don't think that it makes it worse, and I was planning on fixing that one in a separate PR.

Still, I get the concern about stuff going wrong here--it can be dang hard to get it right, and very easy to miss something. Assuming that this is not dead in the water, I fully expect that there will be some back and forth discussions as the ideas gets explored and hopefully it stands up to attempts to rip it to shreds--it is important to get it right.

peter-mitsis replied:
Now that it is day time and I managed to get _some_ sleep last night, I have been taking (yet) another look and I need to take a closer look at the following two routines (and who calls them)--z_set_prio() and k_thread_deadline_set().

The routine z_set_prio() is called by z_thread_priority_set(), and that is not looking like it is quite right for SMP--which in turn would affect the behaviour here. As for k_thread_deadline_set(), it seems to me that changing the deadline such that it results in a lower effective priority should result in a reschedule on potentially more than 1 core as there may now be a thread in the ready queue that  should take precedence.

peter-mitsis commented:
On the topic of testing, I've been trying to figure out how to do that aside from running the regression tests.  I did write a small hack test to count the number of IPIs being delivered to each core under a couple of scenarios to demonstrate to myself during development that fewer IPIs were being processed, but just counting IPIs is not the most practical way to test this--particularly since at present some architectures support targeted IPIs, while others do not. If anyone has some ideas on how I an automated test could be designed for this, I am all ears.

I like the idea about an arch flag like CONFIG_ARCH_HAS_DIRECTED_IPI. I'll see what I can do about that.

marc-hb commented:
On this (lack of) testing topic, SOF just completed a nightmarish, "Big Bang" Zephyr integration nightmare:

- https://github.com/thesofproject/sof/issues/8818
- https://github.com/orgs/thesofproject/discussions/8909

As often, SMP was one of the star actors.

marc-hb commented on kernel/sched.c:
```diff
@@ -397,15 +399,43 @@ static void move_thread_to_end_of_prio_q(struct k_thread *thread)
+	unsigned int num_cpus = arch_num_cpus();
+	uint32_t  id = arch_proc_id();
+
+	for (uint32_t i = 0; i < num_cpus; i++) {
```
Nit: inconsistent mix of `unsigned` and `uint32_t` to number CPUs.

peter-mitsis replied:
Fixed.

marc-hb commented:
> On this (lack of) testing topic, SOF just completed a "Big Bang" Zephyr integration nightmare

I spoke too fast; we still see a relatively frequent PANIC:
- #69807

evgeniy-paltsev commented on soc/intel/intel_adsp/cavs/multiprocessing.c:
```diff
@@ -121,13 +121,15 @@ void soc_start_core(int cpu_num)
 
-	for (int c = 0; c < num_cpus; c++) {
-		if (c != curr && soc_cpus_active[c]) {
+	for (int c = 0; c < num_cpus; c++, bit <<= 1) {
+		if ((c != curr) && soc_cpus_active[c] &&
+		    ((cpu_bitmap & bit) != 0)) {
 			IDC[curr].core[c].itc = BIT(31);
 		}
 	}
```
Why so complicated?

```suggestion
	for (int c = 0; c < num_cpus; c++) {
		if ((c != curr) && soc_cpus_active[c] && ((cpu_bitmap & BIT(c)) != 0)) {
			IDC[curr].core[c].itc = BIT(31);
		}
	}
```

Ditto in other places.

peter-mitsis replied:
> Why so complicated?

Simplifying ... :)

evgeniy-paltsev commented on kernel/sched.c:
```diff
@@ -458,11 +488,10 @@ static void slice_timeout(struct _timeout *t)
 	 */
 	if (IS_ENABLED(CONFIG_SMP) && cpu != _current_cpu->id) {
-		flag_ipi();
+		flag_ipi(1 << cpu);
```
```suggestion
		flag_ipi(BIT(cpu));
```

TaiJuWu commented:
Just a question.

The ipi will pending here until the cpu is idle and then do it, like arm architecture?

peter-mitsis commented:
I was out of town last week, so I am somewhat late in replying to this.

> The ipi will pending here until the cpu is idle and then have time to do it, like arm architecture?

If I follow your question correctly, then no. My understanding is that the IPI work can conceptually be broken up into three (3) parts.

1. Internally track that IPIs will need to be sent -- see _flag_ipi()_. 
2. Send the IPIs if any need to be sent-- see _arch_sched_ipi()_. The IPIs are sent at scheduling points after a thread has been made ready.
3. Process the IPIs -- see _z_sched_ipi()_. This is done in response to the IPI when the CPU receives it. Although the CPU might have been idle when it received it, that is not a prerequisite. That IPI may interrupt it at almost any time.

peter-mitsis committed:
tests: Add IPI_OPTIMIZE tests

Adds several tests to verify that IPIs can be appropriately
targeted to specific CPUs.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis commented:
The latest set of changes has tried to incorporate much of the previous feedback. Hopefully I have not missed any.

The core of the idea remains intact--we can track which CPUs do not need an IPI. This is generally determined by comparing the priority of the thread (and possibly its deadline) against the currently executing thread on another CPU.

This latest version has a few changes that the previous had missed. Notably ...

1. Tests -- see tests/kernel/ipi_optimize. We engineer a number of scenarios to check that IPIs are being issued and detected as expected.
2. Kconfig options: CONFIG_IPI_OPTIMIZE and CONFIG_ARCH_HAS_DIRECTED_IPIs. The former allows for the kernel to track where the IPIs need to be sent while the latter is used to indicate that which architectures can leverage this information.
3. Deadlines -- Setting them is now a schedule point. Also, they are now factored in a consideration as to whether IPIs are needed.

peter-mitsis commented:
Rebased to account for some recent kernel reorganization.

evgeniy-paltsev commented on kernel/ipi.c:
```diff
@@ -13,15 +13,47 @@ extern void z_trace_sched_ipi(void);
+		}
+
+		cpu_thread = _kernel.cpus[i].current;
+		if ((cpu_thread != NULL) &&
+		    (z_sched_prio_cmp(cpu_thread, thread) < 0)) {
```
Zephyr allows code lines [up to 100 symbols](https://docs.zephyrproject.org/latest/contribute/guidelines.html#coding-style). So let's not make code less readable :)



```suggestion
		if ((cpu_thread != NULL) && (z_sched_prio_cmp(cpu_thread, thread) < 0)) {
```

peter-mitsis commented on kernel/sched.c:
```diff
@@ -1063,22 +1095,65 @@ void z_impl_k_thread_deadline_set(k_tid_t tid, int deadline)
+			 * thread is moved ahead so it arrives earlier.
+			 */
+
+			if (((int)((uint32_t)old_dl - (uint32_t)newdl)) > 0) {
```
Can do.

peter-mitsis commented:
Pushed changes to address @evgeniy-paltsev comments about encapsulating the deadline comparisons.

andyross commented on tests/kernel/ipi_optimize/src/main.c:
```diff
@@ -0,0 +1,741 @@
+ * Verify that waking a thread whose priority is lower than any other
+ * currently executing thread does not result in any IPIs being sent.
+ */
+ZTEST(ipi, test_low_thread_wakes_no_ipis)
```
Seems like a *lot* of boilerplate commonality between these tests.  Can't you refactor such that the thread spawning and IPI trace collection is all common, and just takes a config for "what priorities to use" and "what CPUs should get IPIs"?

peter-mitsis replied:
I'll see what I can do. :)

andyross commented on kernel/sched.c:
```diff
@@ -767,14 +767,44 @@ bool z_thread_prio_set(struct k_thread *thread, int prio)
 		need_sched = z_is_thread_ready(thread);
 
 		if (need_sched) {
+#if defined(CONFIG_SMP)
```
Style: "#ifdef CONFIG_..." is more idiomatic, "#if CONFIG_..." works too.  The use of defined() should be limited to expressions that need to compute boolean expressions, and even then is harder to read.

peter-mitsis replied:
Will update.

andyross commented on kernel/sched.c:
```diff
@@ -767,14 +767,44 @@ bool z_thread_prio_set(struct k_thread *thread, int prio)
+
+			old_prio = thread->base.prio;
+#endif /* CONFIG_SMP */
 			/* Don't requeue on SMP if it's the running thread */
```
This comment is left dangling.  This new logic is doing a LOT more than just checking for _current.

peter-mitsis replied:
Oops. Will fix.

andyross commented on kernel/sched.c:
```diff
@@ -767,14 +767,44 @@ bool z_thread_prio_set(struct k_thread *thread, int prio)
 				dequeue_thread(thread);
 				thread->base.prio = prio;
 				queue_thread(thread);
+#if defined(CONFIG_SMP)
```
Can we clean up the #ifery here? This is sort of hard to read (i.e. an "#if CONFIG_SMP" inside an if() expression whose first term is "!IS_ENABLED(CONFIG_SMP)".  Pick one convention, don't mix them.  Maybe compute old_prio in such a way as to have it drop out in the optimizer, make flag_ipi() a noop when !SMP (is it not already?), etc...

andyross replied:
Actually, to be perfectly truthful: the more I look at this logic the less I like it.  Even ignoring the preprocessor-vs-C-logic issues, this is a lot of special-case code added just to handle the case of correctly optimizing IPI delivery for the case of a change in thread priority.   That's just not a performance case in all but the rarest apps; thread priorities are expected to be largely static while they're doing useful work.

andyross commented on kernel/sched.c:
```diff
@@ -767,14 +767,44 @@ bool z_thread_prio_set(struct k_thread *thread, int prio)
+
+				cpu = thread_active_elsewhere(thread);
+				if ((cpu != NULL) && (old_prio < prio)) {
+					flag_ipi();
```
This is the kind of thing that creeps me out about changes like this.  This is going to iterate (!!) over all the CPUs in the scheduler every time any thread anywhere changes priority, putting a cost on all those small operations, just to "optimize" the case where you're "dropping the priority of a thread that happens to be running on another CPU".  Which is a crazy rare edge case, right?

This seems like exactly the circumstance where you'd want the simpler code and a broadcast IPI, no?  All these extra cycles add up over time, and suddenly our lean scheduler isn't so lean.

peter-mitsis replied:
Good point. I got so wrapped up in trimming out IPIs I forgot the larger picture.

peter-mitsis replied:
I've been thinking about this one some more, and the more I think about it, the more I really do think that it is beneficial. I think we are both in agreement that IPIs are needed on the "reschedule is needed" case. Where we primarily differ is the cost/benefit of the IPI calcuation.

While it is true that the cost for the CPU calling z_thread_prio_set() increases slightly, the total cost across the system decreases. That is, we can improve the overall performance at the cost of a slight increase in complexity.

There are three levels of calculations to consider (N shall be the number of CPUs) ...
1. When broadcasting, the total cost is  ...
`total average cost = (N - 1) * (time to service IPI)`
2. When conditionally broadcasting ...
`total average cost = (time to loop over N items) + (probability of crazy rare edge case) * (N - 1) * (time to service IPI)`
3. When targeting, the total average cost is ...
`total average cost = (time to loop over N items) + (probability of crazy rare edge case) * (time to service IPI)`

It is reasonable to expect that the "time to loop over N items" is going to be much lower than the time to service one interrupt (let alone N-1). Furthermore, as the probability decreases, so too does the contribution of the IPI servicing.

andyross commented on kernel/sched.c:
```diff
@@ -1095,23 +1106,65 @@ static inline void z_vrfy_k_thread_priority_set(k_tid_t thread, int prio)
+	key = k_spin_lock(&_sched_spinlock);
+
+	if (z_is_thread_ready(thread)) {
+#if defined(CONFIG_SMP)
```
FWIW I think this change is wrong.   There's no valid usage of EDF scheduling that requires setting a deadline on a running thread.  Deadlines get precomputed in all idioms I'm aware of.  If the app needs it to be a scheduling point the app can synthesize one on its own.

peter-mitsis replied:
I did this purely for completeness from a theory POV. However, I must admit lack of knowledge when it comes to practical usage. If you are certain that there is no valid usage EDF scheduling that requires setting a deadline on a running thread, I am happy to defer to your judgement and undo these EDF changes.

andyross commented on arch/arc/Kconfig:
```diff
@@ -406,4 +406,7 @@ config CMSIS_V2_THREAD_MAX_STACK_SIZE
 	default 2048 if 64BIT
 
+config ARCH_HAS_DIRECTED_IPIS
+	default y
```
Doing this with a select rom the ARCH_whatever kconfig is a little cleaner IMHO.

peter-mitsis replied:
I will see what I can do.

andyross commented on arch/arc/core/smp.c:
```diff
@@ -131,17 +131,19 @@ static void sched_ipi_handler(const void *unused)
 	 * if the target is current core, hardware will ignore it
 	 */
-	unsigned int num_cpus = arch_num_cpus();
+	uint32_t num_cpus = (uint32_t)arch_num_cpus();
```
arch_num_cpus() is a specified API that returns an unsigned int.  Why muck with typing here?

peter-mitsis replied:
Just addressing a previous comment about mixing uint32_t and unsigned int types. I will look again at this. There is probably a cleaner way.

andyross commented on tests/kernel/smp/src/main.c:
```diff
@@ -732,16 +732,19 @@ void z_trace_sched_ipi(void)
 		/* issue a sched ipi to tell other CPU to run thread */
 		sched_ipi_has_called = 0;
-		arch_sched_ipi();
+		arch_sched_ipi((1 << num_cpus) - 1);
```
I'm seeing this expression in a lot of places.  Maybe expose a K_ALL_CPUS_MASK or something instead?  Could share with the cpu_mask APIs.

peter-mitsis replied:
Can do.

andyross commented on kernel/Kconfig.smp:
```diff
@@ -73,6 +73,21 @@ config TRACE_SCHED_IPI
 
+config IPI_OPTIMIZE
+	bool "Optimize IPI delivery"
+	default y
```
Pretty sure you want n here, no?  Most RTOS apps don't do a lot of scheduling and want the smaller code size and O(1) scheduler performance.

peter-mitsis replied:
Oops. This is a remnant from an earlier draft. I thought I had changed it 'n', but apparently not. Thanks for the catch!

andyross reviewed (changes requested):
FWIW I think the second patch should come out.  It's not strictly related, we don't want it IMHO and we should probably have that fight in a separate PR. 

The rest is mostly nitpicking.   One thing I don't see is test coverage of the IPI_OPTIMIZE=n case for the older tests, everything just moves directly to the new scheme and the simpler code is going to rot.   Would be better to set the default =n and e.g. pick an SMP platform and set IPI_OPTIMIZE=y 

Also, some performance data wouldn't be bad to get on this.  The feature is of the form "take small overheads on lots of common operations to avoid large overheads on uncommon ones".  And IMHO that's a tradeoff that is really easy to make in the wrong direction.

andyross commented on kernel/sched.c:
```diff
@@ -34,12 +34,34 @@ static void update_cache(int preempt_ok);
+ * = 0 -> d1 and d2 are the same deadline
+ * < 0 -> d1 occurs after d2
+ */
+int32_t z_sched_deadline_cmp(uint32_t d1, uint32_t d2)
```
Should be `static ALWAYS_INLINE`, no?   This logic is too subtle to express as an "API".  Factoring it out to eliminate duplication is fine, but let's load as few footguns as possible.

peter-mitsis replied:
Will do.

andyross commented on kernel/ipi.c:
```diff
@@ -13,15 +13,47 @@ extern void z_trace_sched_ipi(void);
+		}
+
+		cpu_thread = _kernel.cpus[i].current;
+		if ((cpu_thread != NULL) &&
```
FWIW, this logic still isn't quite right, though it definitely seems to be wrong in the right/conservative direction: it will flag IPIs in the case where the tested thread is higher priority than the foreign thread always, even if the thread can't run on the other CPU because it's running a cooperative thread and/or disallowed via cpu_mask.

peter-mitsis replied:
If I am following you correctly, at this point in time, that is intentional. The overall intent was not to provide a perfect prediction of what CPUs need an IPI, but to determine in as cheap a manner as practical only if it _might_ need one. Receiving an unnecessary IPI does impact performance (which we are trying to improve), but ought not to impact correctness of scheduling and execution.

peter-mitsis replied:
I was originally thinking of refining this check at a later time, but it looks like it should be straight forward (I was originally worried that there might possibly be some places where we were not protecting things as expected). However, those fears appeared to be unfounded.

One implication is that I have to redo the test suite as CPUs running a cooperative thread will no longer receive an IPI due to a thread being made ready.

peter-mitsis commented on kernel/Kconfig.smp:
```diff
@@ -73,6 +73,21 @@ config TRACE_SCHED_IPI
+	  reduce the number of IPIs (and consequently ISRs) processed by the
+	  system. If not, the only benefit would be to not issue any IPIs if
+	  the newly readied thread is of lower priority than all the threads
+	  currently executing on other CPUs.
```
Can do.

peter-mitsis commented on tests/kernel/ipi_optimize/src/main.c:
```diff
@@ -0,0 +1,741 @@
+		k_thread_create(&thread[i], stack[i], STACK_SIZE,
+				busy_thread_entry,
+				NULL, NULL, NULL,
+				priority - 1, 0, K_NO_WAIT);
```
Throughout the test suite, I have followed a pattern with the busy threads to help engineer a specific circumstance that we use to reliably test this feature. My aim was to guarantee which threads were running on each of the CPUs so as to better model, control and predict where the IPIs will go. My thinking was that a board has the potential to automatically enable drivers or features that create their own threads--threads that could interfere with the test. Manipulating the system to have a cooperative thread on each CPU performing a busy loop allows us to sidestep that concern and better know/predict what each CPU is doing. Admittedly those same hypothetical drivers/features could in theory still have ISRs which could interfere with the test, but I figured that I would draw the line somewhere.

peter-mitsis commented on tests/kernel/ipi_optimize/src/main.c:
```diff
@@ -0,0 +1,741 @@
+
+	key = k_spin_lock(&ipilock);
+	ipi_count[_current_cpu->id]++;
+	k_spin_unlock(&ipilock, key);
```
I did consider that. The reason I did not go that route was because when getting and clearing the IPI counts via get_ipi_counts() and clear_ipi_counts() respectively, I wanted to ensure that a snapshot of all of them were obtained at the same time. Atomics would be useful for each element, but unless I have missed something not for the whole array.

peter-mitsis commented on tests/kernel/ipi_optimize/src/main.c:
```diff
@@ -0,0 +1,741 @@
+				priority - 1, 0, K_NO_WAIT);
+	}
+
+	/* Align to tick boundary to minimize probability of timer ISRs */
```
I am not so sure that I will still be able to do timed busy waits if the timer driver is disabled. However, I will still explore this approach a little more.

For what it is worth the two largest concerns I had with how I structured the test centred around time--timer interrupts messing with results, and host system load delaying the delivery/processing of IPIs on QEMU systems to such an extent that the busy wait delays would be insufficient.

peter-mitsis replied:
Disabling the timer via SYS_CLOCK_EXISTS is not practical at this point in time as the ztest infrastructure has calls to k_cycle_get_32() which for at least some platforms requires the system clock. Unless I have missed it, we do not currently have a generic API for disabling the timer interrupt. That leaves decreasing the system tick rate. Setting that to 50 or so ought to be low enough to reasonably minimize the probability of getting a spurious timer interrupt.

Admittedly that still leaves some room for possible timing errors related to host system load, but the zassert_true() checks used to verify system states should provide enough feedback to hopefully determine where/why the failure occurred without the need to wait for the test framework to timeout. If it turns out that the current busy waits are too short, we can always increase them at a later date.

evgeniy-paltsev replied:
@peter-mitsis,

just for note regarding the SYS_CLOCK_EXISTS=n - we still can create test which is not ztest-based - so k_cycle_get_32 dependency can be easily avoided. There are such tests in existing test suite, i.e: `tests/misc/print_format`

peter-mitsis replied:
I can get so used to working with certain patterns and frameworks for things (such as using ZTEST for testing), that I can forget that we don't always have to use them. That being said, I still need the system clock for busy waits. Fortunately, the TICKLESS_KERNEL=y helps further minimise the likelihood of timer ISRs.

peter-mitsis commented on tests/kernel/ipi_optimize/src/main.c:
```diff
@@ -0,0 +1,741 @@
+		}
+
+		clear_ipi_counts();
+		arch_sched_ipi(BIT(i));
```
Will do.

peter-mitsis replied:
I did some further digging into this and observed that although test_arch_sched_ipi() will execute on CPU0, most of the time it  executes on another CPU. With this natural variation between runs, I am presently inclined to leave it as is.

marc-hb commented on kernel/sched.c:
```diff
@@ -34,12 +34,34 @@ static void update_cache(int preempt_ok);
+	 * Using unsigned types and casting to signed isn't perfect, but at
+	 * least reduces this from UB on overflow to impdef.
+	 */
+	return (int32_t) (d2 - d1);
```
At the risk of being too naive with C integers once more:

```C
return d1 < d2 ? d2 - d1 : - (d1 - d2);
```

Doesn't that avoid `impdef`?

I disassembled the output of clang -O2 (version 15) and on X86_64 the code generated is strictly identical to `return d2-d1;`

peter-mitsis commented on kernel/ipi.c:
```diff
@@ -13,15 +13,47 @@ extern void z_trace_sched_ipi(void);
+#elif (CONFIG_MP_MAX_NUM_CPUS > 1)
+	ARG_UNUSED(thread);
+
+	return (1 << CONFIG_MP_MAX_NUM_CPUS) - 1;  /* Broadcast IPIs */
```
Although I understand the potential inconsistency, I do not think that it is in fact an issue as long as CONIFG_MP_MAX_NUM_CPUS >= arch_num_cpus(). As long as that condition holds the kernel might flag a few IPIs for non-existent cores, but arch_sched_ipi(x) will not send IPIs to non-existent cores.

peter-mitsis commented:
I realized this morning that I had a potential bug in the previous version of ipi_mask_create(). When I had added checks to not send an IPI if the target CPU's current thread was not preemptible, I had failed to consider if the new thread was a meta IRQ thread. That should be fixed now.

peter-mitsis commented:
Rebased to resolve merge conflicts.

andyross commented on arch/riscv/Kconfig:
```diff
@@ -391,6 +391,9 @@ config GEN_IRQ_VECTOR_TABLE
 	default y if !SMP
 
+config ARCH_HAS_DIRECTED_IPIS
+	default y
```
Nitpick: this wants to be a select, not a default.  It's not a tunable an app is expected to change, the architecture always provides the API

peter-mitsis replied:
That was a leftover artefact from previous development revisions. Removed.

andyross commented on arch/arm/core/cortex_a_r/smp.c:
```diff
@@ -240,8 +240,10 @@ void sched_ipi_handler(const void *unused)
 {
+	ARG_UNUSED(cpu_bitmap);
+
 	broadcast_ipi(SGI_SCHED_IPI);
```
The implementation of this function has a loop over CPUs doing gic_raise_sgi() already, so surely this arch can do this.  If you don't want to muck with ARM code directly, maybe @carlocaione @npitre et. al. might have input?   But for sure we want to get ducks in a row before merge, either with code in the PR or an active issue to add it.

peter-mitsis replied:
Can do.

andyross commented on arch/arm64/core/smp.c:
```diff
@@ -210,8 +210,10 @@ void sched_ipi_handler(const void *unused)
 {
+	ARG_UNUSED(cpu_bitmap);
+
 	broadcast_ipi(SGI_SCHED_IPI);
```
Ditto from above.  (And an unrelated whine: the GIC code looks basically identical between Cortex A/R and arm64.  Why the cut/paste?)

peter-mitsis replied:
Can do.

andyross reviewed (commented):
Some more notes.  The big points are:

* I think I'm getting an ick about changing arch_sched_ipi() given that some platforms are stubbing the new scheme with a broadcast.  If we can't get everyone to work, let's leave the old broadcast code in place (maybe with a rename, like arch_ipi_broadcast() or something) so no one gets fooled about a feature that doesn't exist.

* Alternatively, it really looks like the existing architectures all have capability to do this, so if they all can then let's just do it and make this a requirement (in which case the API change is just fine as it doesn't imply a capability that doesn't exist).

andyross commented:
One final note that occurs to me: since this is an optimization with unclear tradeoffs, maybe it would be good to add a simple benchmark for it, like a whitebox controlled by a kconfig that can count the fraction of IPIs that result in a context switch (vs. the "useless" ones) and then run it over tests/kernel/* or whatever?

peter-mitsis commented on kernel/sched.c:
```diff
@@ -349,9 +349,9 @@ static void update_cache(int preempt_ok)
-	 * constant time, but this is fine for now.
+	/* Returns pointer to _cpu if the thread is currently running on
+	 * another CPU. There are more scalable designs to answer thisu
+	 * question in constant time, but this is fine for now.
```
Can do.

peter-mitsis commented on arch/x86/core/intel64/smp.c:
```diff
@@ -34,8 +34,10 @@ int arch_smp_init(void)
  * assumes the use of a local APIC (but there's no other mechanism).
  */
-void arch_sched_ipi(void)
+void arch_sched_ipi(uint32_t cpu_bitmap)
```
Given the difference in need between the two, I would like to then propose renaming arch_sched_ipi() to arch_sched_broadcast_ipi() and if directed IPIs are supported add in arch_sched_directed_ipi().

peter-mitsis commented:
Breaking the conversation flow as github will not allow me to respond directly ...

> Again, x86 should be relatively straightforward (and in Intel's wheelhouse, heh).
-   I would prefer to file an enhancement and come back to this later. Hardware is one of my (many) weak areas and going through the documentation and cross referencing with our existing code base is raising a significant number of questions related to the APIC ID (which seems to be programmable). I would hope that the lack of this enhancement on x86_64 would not be a blocking issue if it is tracked via a github issue.

> ... maybe it would be good to add a simple benchmark for it... 
- As to the benchmarking, I am happy to add that to my list of benchmarks to develop, and can create a github issue to track that.

peter-mitsis commented:
Created issues #72170 and #72171 to track the addition of x86 directed IPI support and benchmarking the IPI optimization.

andyross commented on arch/Kconfig:
```diff
@@ -24,6 +24,7 @@ config ARC
 	imply XIP
 	select ARCH_HAS_THREAD_LOCAL_STORAGE
 	select ARCH_SUPPORTS_ROM_START
+	select ARCH_HAS_DIRECTED_IPIS if SMP
```
Nit: this feature doesn't require SMP=y, which is a higher level abstraction in the kernel, not the arch layer.  It's enough to just declare that you can do it, the kernel will then use the API to implement whatever it wants, which may or may not be an SMP kernel.  See tests/kernel/mp for examples of exercising the MP layer of the arch code without building an SMP kernel.

peter-mitsis replied:
I can fix that in each of the places.

andyross commented on kernel/ipi.c:
```diff
@@ -13,15 +13,62 @@ extern void z_trace_sched_ipi(void);
+#else
+	ARG_UNUSED(thread);
+
+	return 0;                                  /* No IPIs needed */
```
Style nit: these two cases can be IMHO better expressed as simple if() expressions (and likely as early returns) than big #if blocks.  The compiler will remove the dead code if the kconfigs don't match, and I don't see any use of fields or APIs here that wouldn't work generically.  Zephyr is getting really bad about complicated preprocessor code blocks these days.

andyross commented on kernel/ipi.c:
```diff
@@ -13,15 +13,62 @@ extern void z_trace_sched_ipi(void);
+#endif
+		    (thread_is_metairq(thread) ||
+		     (thread_is_preemptible(cpu_thread) &&
+		      (z_sched_prio_cmp(cpu_thread, thread) < 0)))) {
```
Style nit:  please no #if around subexpresssions.  Ew :).  Abstract the cpu_mask test into a predicate function like the other steps here and put the preprocessor logic there.

Performance nit: I'd reorder this.   First to remove the test for NULL, which should never happen in a live kernel[1].  After that IMHO by far the most common decision case is the simple priority comparison, then followed by preemptible priorities with metairq last (which should be rare in almost all circumstances, given that they block the world apps will work hard to make them quick).  I'd put cpu_mask last too, but there's some argument as whether or not it's a common case is app-dependent; but my argument would be that apps that use cpu_mask a lot (like SOF) wouldn't use this feature anyway as they will be doing coarse scheduling at the app level (e.g. SOF basically has only one thread per core and doesn't do much kernel scheduling).

[1]  Under no circumstances should a CPU receive an IPI before it's been initialized (i.e. everything should be set before it unmasks interrupts for the first time).  Maybe replace with an assert if you're worried about a specific edge case?

peter-mitsis replied:
I am happy to reorder the checks. However, I do think that the test for NULL is necessary. It ensures that we do not do not dereference a NULL pointer. Recall that static threads get initialized (and may start) before the other CPUs are initialized. Uninitialized CPUs have their _current_ field set to NULL. When these threads start, they can flag IPIs which can then get signalled.

andyross replied:
Ah, indeed.  This is the sending case, not the receiving case, and not all the cpus may be online at a given time (intel_adsp can actually shut them down). 

Which actually brings up a good question: do we null-out the current field of an offline CPU anywhere?  I think we might not, in which case this is going to be stuck throwing IPIs at dead CPUs (which is... probably benign in most cases?  I guess it depends on architecture).  We don't really have a single authoritative place to answer\ "is this CPU alive?".  Not something for this PR but almost certainly something on which we'll stumble in the future.

andyross commented on kernel/sched.c:
```diff
@@ -994,9 +1025,12 @@ void z_impl_k_thread_priority_set(k_tid_t thread, int prio)
+			z_reschedule_unlocked();
+		} else {
+			signal_pending_ipi();
+		}
```
Had to dig here trying to figure out what was going on.  IPI logic absolutely doesn't belong up here in k_thread_priority_set().   I think I get what you're trying to do: we had a gate on the reschedule here based on "sched_locked"[1].  But in SMP that's a little wrong, because even if you can't reschedule locally you still need to tell other CPUs that the local CPU has reached a scheduling point.

But the fix there is to remove the sched_locked logic in SMP[2], or arguably entirely (k_sched_lock() is a very rarely used API, I'd expect it's not worth the extra test here).

Basically what you're actually doing here is working around a pre-existing bug.  Let's fix that instead.

[1] Which is a pretty subtle thing.  It's a holdover from way back in the Ben Walsh era that reflects the top bits of the 16 bit priority such that you can decrement it to make it negative and thus create a "maximally low" priority value.

[2] The schedule lock simply doesn't work in SMP absent great care, and in any case this is just a performance optimization.  All the normal priority logic will prevent preemption of the current thread, we don't actually need this code here at all.

peter-mitsis replied:
If I follow both you and the code correctly, the `(_current->base.sched_locked == 0U)` in k_thread_priority_set() exists only for UP platforms that use arch_swap(). Aside--the arch_switch code is both so much cleaner and easier to follow.

Digging in a little more suggests do_swap() is the correct place to signal the IPI. However, the present version of do_swap() is lacking as it only signals any flagged IPIs if a new thread is to be switched in (see `z_requeue_current()`). But since the current thread could be schedule locked, if there any IPIs flagged for signalling even if the current thread is not going to be switched out, then these should still be signalled. For what it is worth, `z_get_next_switch_handle()` does not have this issue.

Updated code coming soon.

peter-mitsis replied:
Oops ... I wrote too soon. I forgot about `z_swap_next_thread()`. It already does the `signal_pending_ipi()`.

andyross commented on kernel/sched.c:
```diff
@@ -720,19 +723,43 @@ void z_unpend_thread(struct k_thread *thread)
+					 * priority of a queued thread.
+					 */
+
+					flag_ipi();
```
Style nit: Can

you

use

fewer

lines?

Seriously it's a one-line thing, I really don't think you need a comment, and certainly not extra whitespace.  Also no need to test SMP as flag_ipi() is already a stub if that's not active.

```c
if (old_prio > prio) {
     flag_ipi();
}
```

peter-mitsis replied:
Man, this comment made me immediately laugh. It still brings a chuckle every time I read it. :) I  can amend this.

andyross commented on kernel/timeslicing.c:
```diff
@@ -58,11 +58,12 @@ static void slice_timeout(struct _timeout *timeout)
 	 */
-	if (IS_ENABLED(CONFIG_SMP) && cpu != _current_cpu->id) {
-		flag_ipi();
+	if (IS_ENABLED(CONFIG_SMP) && (cpu != _current_cpu->id)) {
```
Same nit from elsewhere.  flag_ipi() is already stubbed when CONFIG_SMP=n, no need to test.

peter-mitsis replied:
Can do.

andyross commented on kernel/timeslicing.c:
```diff
@@ -58,11 +58,12 @@ static void slice_timeout(struct _timeout *timeout)
-	 * the specific core, but that's not part of the API yet.
+	 * for a different CPU.  Ideally the underlying architecture
+	 * allows for the IPI needed for CPU <cpu> to be directed
+	 * only to it and not any other CPUs.
```
Why "ideally"?  That's the way it works now.  My comment was probably silly to begin with, but the limitation it was trying to document doesn't exist anymore.  Just remove.  "We need an IPI if we just handled a timeslice expiration for a different CPU" is all that needs be said.

andyross commented on arch/arm64/core/smp.c:
```diff
@@ -180,7 +181,7 @@ void arch_secondary_cpu_init(int cpu_num)
 #ifdef CONFIG_SMP
 
-static void broadcast_ipi(unsigned int ipi)
+static void send_ipi(unsigned int ipi, uint32_t cpu_bitmap)
```
Style nit: the platform maintainers might prefer to see each arch get support added in a separate patch instead of all at once, but that's not really for me to say.

andyross reviewed (commented):
OK, last round of nitpicks and notes I think.  I promise I've gone through the thing in totality at this point.  I'll dismiss the -1, I don't think there's anything here that shouldn't go in; most of my current notes are minor.

We do need to get the ARM/ARC/RISCV folks in to make sure they have a +1 on it before merge.  I guess I'm qualified to OK xtensa changes myself.

And I still note that it's a deep irony given the domain in your email address that the **only** SMP architecture without support is x86.  But if it's just one outlier I guess there's no reason not to merge. :)

andyross reviewed (dismissed):
I'm out of complaints.  But we do need relevant arch folks to +1 too.  @carlocaione @edersondisouza @evgeniy-paltsev @npitre @SgrrZhf seem like obvious candidates

npitre commented:
Someone is currently removing all the `#if CONFIG_*` at the moment.
Might be better not to reintroduce new ones i.e.
`#if CONFIG_ARCH_HAS_DIRECTED_IPIS` should be
`#ifdef CONFIG_ARCH_HAS_DIRECTED_IPIS` (there is at least one occurrence 
in the test code).

marc-hb commented:
> Someone is currently removing all the #if CONFIG_* at the moment.

Interesting, is this policy documented somewhere? Only for CONFIG_*?

PS: can any `CONFIG_*` ever be zero? I doubt it.

npitre commented:
See PR #72088

peter-mitsis commented:
Based on @npitre's comment, I have updated the test to use `#ifdef` instead of `#if`.

peter-mitsis commented on arch/arc/core/smp.c:
```diff
@@ -130,21 +131,27 @@ static void sched_ipi_handler(const void *unused)
 	}
 }
 
+void arch_sched_broadcast_ipi(void)
```
arch_sched_broadcast_ipi() I don't think should be using ALWAYS_INLINE as it is not being called from anywhere else in this file. Although arch_sched_directed_ipi() could, I don't think it would get us too much--probably better to let the compiler determine if arch_sched_broadcast_ipi() inlines arch_sched_directed_ipi() or makes a call out to it.

fkokosinski reviewed (dismissed):
RISC-V changes LGTM.

TaiJuWu commented:
Just curious, is all architecture support IPI?
If an architecture don't offer ipi in hardware layer,  what will happen?

andyross commented:
@TaijuWu we don't technically require IPIs.  If an arch leaves CONFIG_SCHED_IPI_SUPPORTED=n then an SMP kernel fakes it by busy-waiting in the idle loop trying to swap.  That's not really intended to be a production use case, but is helpful during bringup, etc...

TaiJuWu commented:
> @TaiJuWu we don't technically require IPIs. If an arch leaves CONFIG_SCHED_IPI_SUPPORTED=n then an SMP kernel fakes it by busy-waiting in the idle loop trying to swap. That's not really intended to be a production use case, but is helpful during bringup, etc...

Got it, thanks for your answer.

evgeniy-paltsev commented:
@peter-mitsis, I've tested this - the `kernel.ipi_optimize.smp` test hangs on most Synopsys (ARC) smp platforms. I'll need to investigate this.

Here is a test output:
```
Running TESTSUITE ipi
===================================================================
START - test_arch_sched_broadcast_ipi
```

evgeniy-paltsev commented on arch/arc/core/smp.c:
```diff
@@ -130,21 +131,27 @@ static void sched_ipi_handler(const void *unused)
 
 	for (i = 0U; i < num_cpus; i++) {
-		z_arc_connect_ici_generate(i);
+		if ((cpu_bitmap & BIT(i)) != 0) {
+			z_arc_connect_ici_generate(i);
+		}
```
Ok, now we have issue here for ARC as core ID and ARConnect ID doesn't always match. Currently we have it so for SMP ARC nSIM (simulation) platforms due to the way we launch it.

I'll check if it is possible to force the IDs so the would match for nSIM platforms as well.

evgeniy-paltsev replied:
Now it's fixed with https://github.com/zephyrproject-rtos/zephyr/pull/72755

peter-mitsis commented:
Rebased to resolve merge conflicts.

teburd commented:
@lgirdwood @kv2019i @lyakh please take a close look as this relates to your observations of ipc interrupt counts

peter-mitsis commented:
Rebased to resolve a merge conflict.

andyross reviewed (dismissed):
Refresh +1

nashif commented on submanifests/optional.yaml:
```diff
@@ -34,7 +34,7 @@ manifest:
         - optional
     - name: sof
-      revision: c11a3185afbc8e1b2a79916de3dfefaf326d9ad1
+      revision: ac1b774abfe2e8a9f6547f5bc305e499c51f9671
```
point to this d3951659ea4263fdb74b76688587d04d91b2e58c instead.

nashif replied:
@peter-mitsis  ^^

nashif commented:
> but the conflict here not sure what is it about.

ok, that was the esf change...

peter-mitsis commented:
Updated submanifest as per @nashif's request -- nothing else changed.

andyross reviewed (approved):
Again
